### PR TITLE
Only clear written-to locals in ConstProp

### PR DESCRIPTION
**Pull request by cjgillot**
This aims to revert the regression in https://github.com/rust-lang/rust/pull/108872

Clearing all locals at the end of each bb is very costly.
This PR goes back to the original implementation of recording which locals are modified.

### CODE REVIEW

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 1]
```diff
@@ -691,17 +692,30 @@ impl<'tcx> Visitor<'tcx> for ConstPropagator<'_, 'tcx> {
...
+            debug_assert_eq!(
+                self.ecx.machine.can_const_prop[local],
+                ConstPropMode::OnlyInsideOwnBlock
+            );
+            Self::remove_const(&mut self.ecx, local);
+        }
+        self.ecx.machine.written_only_inside_own_block_locals =
+            written_only_inside_own_block_locals;
+
+        #[cfg(debug_assertions)]
```
Comment (Nilstrieb) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs:711]: could you turn this into an `if cfg!(debug_assertions)` instead? it does increase the indentation slightly but it's a lot nicer to work with

**Review by davidtwco — APPROVED** [review]
r=me, resolving nils comment if you want

**Comment by nnethercote** [comment]
Oh dear: #108815 was a big improvement for `keccak` and `cranelift-codegen`. Unfortunately, it was merged between #108872 (which massively regressed those benchmarks) and this PR (which mostly?/entirely? fixes the regression). It will be tricky to disentangle the individual perf effects of the three PRs.

cc @the8472

**Comment by nnethercote** [comment]
Failure while building LLVM:

> could not create temporary file whilst writing archive: no more archived files

Seems like a transient error unrelated to this PR.

@bors retry

**Pull request merged**